# Week 1 Challenge 
Guess the data from a given received signal file and FIR of the channel 

Parameters:
- Constellations:
    - Bytes were mapped to QPSK constellations 2 bits at a time  (00, 01, 11, 10)
- IDFT:
    - size of 1024
    - Cyclic prefix of length 32
    - 2 bit groups were added at the end of the data to complete an integer factor of 511
    - the resulting blocks of 1056 values were concatenated into one long signal


### 0.1 Import Library

In [11]:
import pandas as pd
import numpy as np

### 1.1 Import channel and files

In [44]:
FIR = np.genfromtxt('channel.csv', delimiter=',')

In [45]:
file1 = np.genfromtxt("file1.csv", delimiter=',')

### 1.2 Check the length of the file

In [46]:
len(file1)

1003200

In [32]:
len(file1)/(1024+32)

950.0

### 2. Remove Cyclic Prefix

#### 2.1 First transform the 1D array into 2D array 
    
    where each row is of 1056 long and there is 950 such rows

In [47]:
file1_2d = file1.reshape(-1,1056)

In [48]:
file1_2d.shape

(950, 1056)

#### 2.2 remove cyclic prefix
    Equivalent to delte columns from index 1024 to 1055(last)

In [58]:
remove = np.arange(32) + 1024
remove

array([1024, 1025, 1026, 1027, 1028, 1029, 1030, 1031, 1032, 1033, 1034,
       1035, 1036, 1037, 1038, 1039, 1040, 1041, 1042, 1043, 1044, 1045,
       1046, 1047, 1048, 1049, 1050, 1051, 1052, 1053, 1054, 1055])

In [59]:
file1_without_cp = np.delete(file1_2d,remove,axis=1)

In [60]:
file1_without_cp.shape

(950, 1024)

### 3. Apply DFT

In [61]:
file1_dft = np.array( [np.fft.fft(file1_without_cp[i,:]) for i in range(file1_without_cp.shape[0]) ] ) 

In [62]:
file1_dft.shape

(950, 1024)

#### 4. Apply channel equalization 

In [63]:
H = np.fft.fft(FIR, 1024) # impulse response of the channel

In [75]:
Equalized_output = file1_dft / H